### DISCUSSION TITLE
How to detect coloured/black background?

### DISCUSSION THREAD

**Discussion by keto33** [ImageMagick/ImageMagick#3004] [Help]
I convert colour images to Black & White as

`magick JPG -type Grayscale PNG`

When the background is coloured, say a white text on a red sheet, the Black&White image is naturally a white text on a black background. I wish to keep the background white in all cases. Thus, I need to invert if the resulting background is black.

I can think of two approaches (which are not perfect but can work):

- Checking if the dominant colour is black.
- Checking if a complete row of pixels is black.

However, I do not know how to apply these conditions in the magick command.

I use ImageMagick 7.0.10-16 on Linux.

**Comment by snibgo** [comment]
Text takes a small proportion of a page. So if you have black text on a white background, the average lightness will be high. And white text on a black background will have a very low average. If you know your image is one of those, then testing is easy.

But if you might have white (or black) text on a gray background, do "-auto-level" first. This will increase contrast so the image has at least one black and one white pixel.

**Comment by keto33** [comment]
As you can guess, the input will be a variety of different images, but it is reasonable to judge the lightness.

Could you please elaborate on the easy testing you mentioned?
